pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,412 @@
+	progressing, err := c.sync(ctx, syncCtx)
+	if err != nil {
+		return err
+	}
+
+	// nil == keep the current progressing status
+	if progressing == nil {
+		return nil
+	}
+
+	var progressingText string
+	progressingStatus := operatorv1.ConditionFalse
+	progressingReason := "CertificateAvailable"
+	if *progressing {
+		progressingText = "waiting for CSR to be signed"
+		progressingStatus = operatorv1.ConditionTrue
+		progressingReason = "WaitingForCertificate"
+	}
+
+	_, _, statusErr := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "AuthenticatorCSRProgressing",
+			Status:  progressingStatus,
+			Reason:  progressingReason,
+			Message: progressingText,
+		}))
+	return statusErr
+}
+
+func (c *webhookAuthenticatorCertRequester) sync(ctx context.Context, syncCtx factory.SyncContext) (progressing *bool, err error) {
+	key, secretComplete, shouldResubmit, err := c.checkKeyCertSecret(ctx, syncCtx.Recorder())
+	if err != nil {
+		return nil, err
+	}
+
+	if secretComplete {
+		return pbool(false), err
+	}
+
+	if !shouldResubmit {
+		var cert []byte
+		cert, shouldResubmit, err = c.handleCSR(key, syncCtx.Recorder())
+		if err != nil {
+			return nil, err
+		}
+
+		if cert != nil {
+			c.attachCertToSecret(cert, syncCtx.Recorder())
+			// we've got the cert, stop progressing
+			return pbool(false), nil
+		}
+
+		if !shouldResubmit { // keep on waiting for the current CSR
+			return pbool(true), nil
+		}
+
+	}
+
+	// either we did not have a key or the CSR failed
+	if shouldResubmit {
+		if err := c.resubmitCSR(ctx, key); err != nil {
+			return nil, err
+		}
+		return pbool(true), nil
+	}
+
+	return pbool(false), nil
+}
+
+func (c *webhookAuthenticatorCertRequester) checkKeyCertSecret(ctx context.Context, recorder events.Recorder) (key *rsa.PrivateKey, secretComplete, shouldResubmit bool, err error) {
+	key, cert, err := c.getKeyCertFromSecret()
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, false, false, err
+	}
+
+	if key != nil && cert != nil {
+		// check that the certificate is not expiring
+		certValidityDuration := cert.NotAfter.Sub(cert.NotBefore)
+		if time.Now().Before(cert.NotBefore.Add(time.Duration(0.8 * float64(certValidityDuration)))) {
+			return key, true, false, nil
+		}
+		// set key == nil to trigger new key generation
+		key = nil
+
+		expiryMsg := fmt.Sprintf("the authenticator's certificate is almost expired (%s), will schedule a new CSR creation", cert.NotAfter.String())
+		recorder.Eventf("ClientCertExpiringSoon", expiryMsg)
+		klog.V(5).Infof(expiryMsg)
+	}
+
+	if key != nil {
+		return key, false, false, err
+	}
+
+	// either the key has always been nil, or the certificate is expiring
+	key, err = createSecretWithNewKey(c.secretClient, recorder)
+	return key, false, true, err
+
+}
+
+func (c *webhookAuthenticatorCertRequester) handleCSR(key *rsa.PrivateKey, recorder events.Recorder) (cert []byte, shouldResubmit bool, err error) {
```

Comment:
what does "handle" mean? Create a cert for the key?

Reply:
process CSR's conditions, grab the certificate if it exists